cloudnet/src/main/java/de/dytanic/cloudnet/template/install/ServiceVersionProvider.java

```diff
@@ -0,0 +1,322 @@
+            JsonDocument document = new JsonDocument().read(inputStream);
+
+            if (document.contains("versions")) {
+                Collection<ServiceVersionType> versions = document.get("versions", TypeToken.getParameterized(Collection.class, ServiceVersionType.class).getType());
+
+                for (ServiceVersionType serviceVersionType : versions) {
+                    this.registerServiceVersionType(serviceVersionType);
+                }
+
+                success = true;
+            }
+        }
+
+        connection.disconnect();
+        if (success) {
+            return true;
+        }
+
+        for (ServiceVersionType serviceVersionType : this.createDefaultVersionTypes()) {
+            this.registerServiceVersionType(serviceVersionType);
+        }
+
+        return false;
+    }
+
+    public void registerServiceVersionType(ServiceVersionType serviceVersionType) {
+        this.serviceVersionTypes.put(serviceVersionType.getName().toLowerCase(), serviceVersionType);
+    }
+
+    public Optional<ServiceVersionType> getServiceVersionType(String name) {
+        return Optional.ofNullable(this.serviceVersionTypes.get(name.toLowerCase()));
+    }
+
+    public boolean canBuildVersion(ServiceVersion serviceVersion, ServiceVersionType.InstallerType installerType) {
+        return !installerType.requiresSpecificJavaVersionToExecute() || serviceVersion.canRunOn(JavaVersion.getRuntimeVersion());
+    }
+
+    public void installServiceVersion(ServiceVersionType serviceVersionType, ServiceVersion serviceVersion, ITemplateStorage storage, ServiceTemplate serviceTemplate) {
+        if (!this.canBuildVersion(serviceVersion, serviceVersionType.getInstallerType())) {
+            throw new IllegalStateException("Cannot run " + serviceVersionType.getName() + "-" + serviceVersion.getName() + "#" + serviceVersionType.getInstallerType() + " on " + JavaVersion.getRuntimeVersion().getName());
+        }
+
+        ServiceVersionInstaller installer = this.installers.get(serviceVersionType.getInstallerType());
+
+        if (installer == null) {
+            throw new IllegalArgumentException("Installer for type " + serviceVersionType.getInstallerType() + " not found");
+        }
+
+        if (!storage.has(serviceTemplate)) {
+            storage.create(serviceTemplate);
+        }
+
+        Path workingDirectory = Paths.get(System.getProperty("cloudnet.tempDir.build", "temp/build"), UUID.randomUUID().toString());
+        try (OutputStream outputStream = storage.newOutputStream(serviceTemplate, serviceVersionType.getTargetEnvironment().getName() + ".jar")) {
+            Files.createDirectories(workingDirectory);
+
+            installer.install(serviceVersion, workingDirectory, outputStream);
+
+            FileUtils.delete(workingDirectory.toFile());
+        } catch (IOException exception) {
+            exception.printStackTrace();
+        }
+    }
+
+    public Collection<ServiceVersionType> createDefaultVersionTypes() {
```

Comment:
We could also put the default file in the resources folder, then we dont need this code

Reply:
Yes